SignalServiceKit/src/Storage/OWSStorage.m

```diff
@@ -373,18 +381,36 @@ + (BOOL)isStorageReady
 
 - (BOOL)tryToLoadDatabase
 {
-
-    // We determine the database password first, since a side effect of
+    // We determine the database password, salt and key spec first, since a side effect of
     // this can be deleting any existing database file (if we're recovering
     // from a corrupt keychain).
+    //
+    // Although we don't use databasePassword or databaseSalt in this method,
+    // we use their accessors to ensure that all three exist in the keychain
+    // and can be loaded or that we reset the database & keychain.
```

Comment:
This is kind of important.  Basically, I decided that it would be simplest and cleanest to persist all three values in the keychain (password, salt, key spec) and create them in lockstep so all three should be there or none should be there, even though we only use the key spec when opening the database.

* This allows us to use a simple test for keychain availability - is the password available? - that works for legacy users too.
* This might help with debugging.